### ISSUE TITLE
[Scoper] Scoping autoloaded files result in an autoload conflict

### ISSUE THREAD

**Issue by Smitsel** [humbug/php-scoper#298]
Edit to avoid to summarise the state of this issue.

When scoping a package which has autoloaded files, the scoped files will keep the same hash. As a result if your current application/library is using the scoped package and has files with the same hash, only one file will be loaded which results in undesirable behaviours.

It appears the hash strategy is not something we can change composer/composer#7942 and changing the package name of the scoped packages has undesirable side effects so it is not a solution either.

@jarrodbell https://github.com/humbug/php-scoper/issues/298#issuecomment-678089152 and @ondrejmirtes https://github.com/humbug/php-scoper/issues/298#issuecomment-683425709 made it work by changing it after the composer dump process as part of the scoping build process.

The discussion is carrying from that point to figure out if there is a way PHP-Scoper could fix it by itself without further actions from the user.

<hr/>

## Bug report

| Question    | Answer
| ------------| ---------------
| Box version | PHP Scoper version 0.11.4 2018-11-13 08:49:16 UTC
| PHP version | PHP 7.1.22
| Platform with version | Ubuntu 16.04.5 LTS
| Github Repo | https://github.com/mollie/mollie-api-php

We are using php-scoper in our client to generate a release for integrators to be used in for example Wordpress. When it is used there, there are often more packages with Guzzle via composer.

When this happens they will get the exception:
`Uncaught Error: Call to undefined function GuzzleHttp\choose_handler()`

This is because the corresponding `guzzlehttp/guzzle/src/functions_include.php` is not loaded for the second implementation due to the composer file require hash being the same.

Below there is an example of our package being ran alongside another guzzle package. 
This is a var_dump inside the composer_real.php where the functions_include is being required if the file hash is not already loaded. 

```
array(3) {
  ["c964ee0ededf28c96ebd9db5099ef910"]=>
  string(99) "{path_to_site}/vendor/composer/../guzzlehttp/promises/src/functions_include.php"
  ["a0edc8309cc5e1d60e3047b5df6b7052"]=>
  string(95) "{path_to_site}/vendor/composer/../guzzlehttp/psr7/src/functions_include.php"
  ["37a3dc5111fe8f707ab4c132ef1dbc62"]=>
  string(97) "{path_to_site}/vendor/composer/../guzzlehttp/guzzle/src/functions_include.php"
}
array(3) {
  ["c964ee0ededf28c96ebd9db5099ef910"]=>
  string(106) "{path_to_site}/guzzle/vendor/composer/../guzzlehttp/promises/src/functions_include.php"
  ["a0edc8309cc5e1d60e3047b5df6b7052"]=>
  string(102) "{path_to_site}/guzzle/vendor/composer/../guzzlehttp/psr7/src/functions_include.php"
  ["37a3dc5111fe8f707ab4c132ef1dbc62"]=>
  string(104) "{path_to_site}k/guzzle/vendor/composer/../guzzlehttp/guzzle/src/functions_include.php"
}
```

So this causes only the first implementation it's `choose_handler` function to be available.

I'm not really sure if this bug should be reported here, but i'm curious to your expert opinion about this. Might it be worth it to also scope the files that outta be required by composer? To avoid some of them not being loaded?  As manipulating the hash in above example would fix the problem.

**Comment by theofidry** [comment]
I admit I'm a bit confused.

>Uncaught Error: Call to undefined function GuzzleHttp\choose_handler()

Since it is not a function from the global namespace, unless explicitly configured otherwise, I would expect this function to be scoped. If it is the case, then it means somewhere a piece of code has not been scoped properly and is still using the old reference (which _can_ be an expected case. Unfortunately due to PHP nature PHP-Scoper won't be able to be perfect and will sometimes need some help via the [patchers](https://github.com/humbug/php-scoper#patchers)).

>This is because the corresponding guzzlehttp/guzzle/src/functions_include.php is not loaded for the second implementation due to the composer file require hash being the same.

What I would find weird/confusing there is why would it be a problem for the scoped version and not the unscoped one? Also if the hash file is the same it means the content is the same, so indeed there shouldn't be a need to include it twice should it?

**Comment by Smitsel** [comment]
Yes, i've done a lot of testing with it.

I've tried adding even more code to the `functions_include`.

Does not change a thing, unfortunately.

**Comment by theofidry** [comment]
I think it would be best to try to narrow down the issue in a reproducible repo to ease the debugging. It may be a bug in Composer or it means Composer is relying on the hash from the `installed.json` files which would be troublesome and means we would need to do something about it for PHP-Scoper to work properly in this case

**Comment by Smitsel** [comment]
I'll create a repo for you 👍

**Comment by Smitsel** [comment]
The repository:
https://github.com/Smitsel/guzzle-scoper-reproduction

If you run into any problems or need any more information just let me know :)

**Comment by theofidry** [comment]
👍 I'll try to take a look ASAP but it's likely gonna take a couple of weeks, this month is quite busy for me :/

**Comment by Smitsel** [comment]
Not a problem. We can hack in a solution if it becomes really pressing. Happy that you are willing to take a look!

**Comment by willemstuursma** [comment]
It's probably here: 
https://github.com/composer/composer/blob/master/src/Composer/Autoload/AutoloadGenerator.php#L903

**Comment by Smitsel** [comment]
Yep, manipulating the package name fixes it indeed.

Very weird choice to do it just based on that. They should also compare content-size or something.

**Comment by sandervanhooft** [comment]
I'd hash the whole package directory, but that may be a bit resource expensive 😄

**Comment by theofidry** [comment]
it's for the dumping the autoloader so I don't think it would be a big deal

**Comment by sandervanhooft** [comment]
Would using `getDistSha1Checksum()` be a solution? It's on the PackageInterface.

**Comment by sandervanhooft** [comment]
https://github.com/composer/composer/blob/19ba2edd5c9e86a5786245e6349e979e5ca380d1/src/Composer/Package/PackageInterface.php#L172

**Comment by theofidry** [comment]
I'm not sure. But I think it's a Composer issue so might be better to create an issue and look for a solution there.

If not achievable, maybe we can look for a workaround in PHP-Scoper but I would like to avoid that if possible.

**Comment by willemstuursma** [comment]
@sandervanhooft no its needs to hash the file contents so it can deal with out situation. 

@theofidry Could you open an issue at Composer? I think it will carry more weight if you do it as the owner / maintainer of the humbug.

**Comment by theofidry** [comment]
So as per the discussion in https://github.com/composer/composer/issues/7942, I changed my mind and I think it makes sense for PHP-Scoper to change the names of the scoped packages.

So I guess a solution would be to also change the packages names in `composer.json` and `installed.json` to solve the issue, maybe by just appending the prefix to the name.

**Comment by patrickjahns** [comment]
I've experimented with this a bit and couldn't find a simple way to solve this within scoper.
Once the package-name is altered in `installed.json` - a autoload dump ( which is recommended after scoping right? ) won't work - as the package name does no longer reflect the path on the file system for the vendor libraries.

For the time being I've went to parsing `autoload_static.php` and `autoload_files.php` and prefix the file hashes in there to trick autoloader into loading it again

But this is then a 3 step process:
1) scoper
2) dumper autoloader
3) fix autoloader

Any ideas how to best move forward here?

**Comment by theofidry** [comment]
I think fixing the autoloader itself is a must avoid: we are jumping in an incredibly more complex solution if we do so.

I'm quite busy at the moment so I can't take a proper look but maybe @naderman has an idea?

@naderman to avoid the whole thread, the issue here is that:

- Guzzle relies on autoloaded files for declaring some functions
- The scoped Guzzle works as expected, except the autoloaded files loading strategy relies on the file name + package name, which might be the same as the non-scoped Guzzle

The result is that one file is loaded (e.g. the non scoped Guzzle one) and the scoped one won't be loaded since has the same hash as the non-scoped one. 

Related Composer issue: https://github.com/composer/composer/issues/7942

**Comment by patrickjahns** [comment]
Understood - however I am not sure how to catch this in scoper - when one is supposed to dump the autoloader after using scoper.

What solved it for me was this simple script"

```
<?php
/**
 * This helper is needed to "trick" composer autoloader to load the prefixed files
 * Otherwise if owncloud/core contains the same libraries ( i.e. guzzle ) it won't
 * load the files, as the file hash is the same and thus composer would think this was already loaded
 *
 * More information also found here: https://github.com/humbug/php-scoper/issues/298
 */
$scoper_path = './build/scoper/vendor/composer';
$static_loader_path = $scoper_path.'/autoload_static.php';
echo "Fixing $static_loader_path \n";
$static_loader = file_get_contents($static_loader_path);
$static_loader = \preg_replace('/\'([A-Za-z0-9]*?)\' => __DIR__ \. (.*?),/', '\'a$1\' => __DIR__ . $2,', $static_loader);
file_put_contents($static_loader_path, $static_loader);
$files_loader_path = $scoper_path.'/autoload_files.php';
echo "Fixing $files_loader_path \n";
$files_loader = file_get_contents($files_loader_path);
$files_loader = \preg_replace('/\'(.*?)\' => (.*?),/', '\'a$1\' => $2,', $files_loader);
file_put_contents($files_loader_path, $files_loader);
```

**Comment by theofidry** [comment]
Maybe a way would be to collect the included scoped files and append them to the dumped autoloader?

**Comment by polevaultweb** [comment]
I'm here with the same problem, I have a WordPress plugin bundling the scoped AWS SDK, but it loads a functions file before another plugin with the SDK. 

@theofidry have you had any further thoughts on this?

@patrickjahns is https://github.com/humbug/php-scoper/issues/298#issuecomment-525700081 still your working solution? Where does that code live?

**Comment by theofidry** [comment]
No I couldn't really check. One would need to check if, when the scoped file is appended to the `scoper-autoload.php` like it is already done for class & function aliases, works. If that's the case then it could be possible to collect those kind of files to append them to the scoper autoloader

**Comment by theofidry** [comment]
Thinking about this again, a possible solution would be to have a `FileCollector` class and:

- when coming across a `installed.json` file:
  - collect all the files listed in `autoload.files` (which can only be files, not directories, if I understood correctly) in `FileCollector`
  - remove all those files from the `installed.json`
- when coming across a `composer.json`:
  - collect all the files listed in `autoload.files` in `FileCollector`
  - remove all those files from `composer.json`
- ignore `composer.lock` files

And then when dumping the PHP-Scoper autoloader we can append all those files there and they should not be in the regular Composer `vendor/composer/autoload_files.php` (even after dumping the autoloader again)

**Comment by pxlrbt** [comment]
It seems like the most simple solutions are:
1) Patching the autoloader files after generation as @patrickjahns [proposed](https://github.com/humbug/php-scoper/issues/298#issuecomment-525700081)
2) Changing the `getFileIdentifier()` function inside composer. It could add the composer package name of the current composer.json - if available - to generate a unique hash across multiple packages. 
```php
$composer = json_decode(file_get_contents('path/to/composer.json));
$rootName = isset($composer->name) ? $composer->name . ':' : '';
md5($rootName . $package->getName() . ':' . $path);
```

Is there anything I'm missing?

**Comment by theofidry** [comment]
> It could add the composer package name of the current composer.json - if available - to generate a unique hash across multiple packages.

This won't be enough: the example here is that the conflicting file is from the same package: one scoped and the other not. So instead it should either be based on the content or base on the file path for example

**Comment by pxlrbt** [comment]
>  the example here is that the conflicting file is from the same package: one scoped and the other not.

The initial issue mentioned WordPress plugins. What I understood - and what is my use case - is that `plugin-a/vendor/guzzlehttp` and `plugin-b/vendor/guzzlehttp` collide since the relative paths to the function files are the same (`./vendors/guzzlehttp/src/functions_include.php`) and so is the md5 hash.

For this general WordPress plugin use case including the root package name (`plugin-a` or `plugin-b`) into the md5 hash would solve this issue.

**Comment by theofidry** [comment]
ha correct, the relative path was not enough hence the suggestion of using a hash of the file content instead somewhere

**Comment by pxlrbt** [comment]
Well, this would be a solution too without relying on the optional package name but adds a bit more overhead. As php-scoper runs before dumping the loader, I think this should be either solved by a composer plugin which hooks into the `dump-autoload` command or integrated into composer core. What's your opinion on this?

**Comment by theofidry** [comment]
Hm based on https://github.com/composer/composer/issues/7942#issuecomment-459409836 using the file content doesn't work either.

So I would first try to change the name of the scoped packages, because if works, it would be a quite robust way to approach this without requiring any change in Composer itself or hacking its autoloader

**Comment by pxlrbt** [comment]
> Hm based on composer/composer#7942 (comment) using the file content doesn't work either.
Should be a combination of hashed file path and hashed file content. 

> So I would first try to change the name of the scoped packages
Well is there a difference between hacking the autoloader hashes and hacking the package names? Changing the package names by php-scoper would require composer.json included into the prefixing, right? In my case I only include vendors files so that the rest of the plugin doesn't get namespaced. I feel like updating the autoloader hashes is the most versatile solution.

**Comment by theofidry** [comment]
If you think you can update the autoloader hashes in the right manner it's fine by me. I just want to avoid to have to hack Composer internals, because that's is going to be super fragile.

>Changing the package names by php-scoper would require composer.json included into the prefixing, right

It would at least require the `installed.json` file yes

**Comment by pxlrbt** [comment]
> It would at least require the installed.json file yes

So, I just tried to change the package name in `installed.json` before running `composer dump-autoload`. But that apparently breaks the package detection.

**Comment by theofidry** [comment]
is it the case if you update the .lock as well?

**Comment by pxlrbt** [comment]
> is it the case if you update the .lock as well?

Yep. Changing .lock file doesn't help. Don't know enough about the internals of composer. At the moment I don't have the time to write a composer plugin so I will stick to the hack that updates the hashes.

Maybe a section about this issue can be added to "Limitations > Composer Autoloader" like:

"PHP-Scoper does not support Composers static file autoloaders. Composer uses the package name for autoloading which is not changed. See #298."

**Comment by theofidry** [comment]
At least until that problem is solved, it definitely deserves an entry in the doc. Would you mind opening a PR for it?

**Comment by pxlrbt** [comment]
Created a PR here: https://github.com/humbug/php-scoper/pull/386

**Comment by veewee** [comment]
For what it's worth : I agree this is a php-scoper issue.
Since the suggested workflow is to run the scoper and next to dump the autoloader.


What I suggest is that: 

1. Run https://github.com/humbug/box/blob/master/src/Composer/ComposerOrchestrator.php as usual
2. Let scoper fix autoload as usual  changed by https://github.com/humbug/php-scoper/blob/master/src/Autoload/ScoperAutoloadGenerator.php
3. Provide a way to hook into a "post autoload dump" in box.json or scoper.inc.php



What I did to solve it in my project is a dynamic post-autoload-dump script:

```php
// add-scripts-to-composer.php

$composerFile = __DIR__.'/composer.json';

$composer = json_decode(file_get_contents($composerFile), true);

$new = file_put_contents(
    $composerFile,
    json_encode(
        array_merge_recursive($composer, ['scripts' => [
            'post-autoload-dump' => [
                'GrumPHP\\Composer\\FixBrokenStaticAutoloader::fix'
            ],
        ]]),
        JSON_PRETTY_PRINT
    )
);
```

This is what the hook looks like (pretty much what is suggested above)

```php
<?php

declare(strict_types=1);

namespace GrumPHP\Composer;

use Composer\Script\Event;

class FixBrokenStaticAutoloader
{
    public static function fix(Event $event)
    {
        $autoloadPath = getcwd().'/vendor/composer';

        $static_loader_path = $autoloadPath.'/autoload_static.php';
        $event->getIO()->write(["Fixing $static_loader_path"]);
        $static_loader = file_get_contents($static_loader_path);
        $static_loader = \preg_replace(
            '/\'([A-Za-z0-9]*?)\' => __DIR__ \. (.*?),/',
            '\'scoped$1\' => __DIR__ . $2,',
            $static_loader
        );
        file_put_contents($static_loader_path, $static_loader);

        $files_loader_path = $autoloadPath.'/autoload_files.php';
        $event->getIO()->write(["Fixing $files_loader_path"]);
        $files_loader = file_get_contents($files_loader_path);
        $files_loader = \preg_replace('/\'(.*?)\' => (.*?),/', '\'scoped$1\' => $2,', $files_loader);
        file_put_contents($files_loader_path, $files_loader);
    }
}

```


So if we could somehow make it possible to hook into the post autoload dump section, we can run whatever fix we think is necessary + you could provide a utitlity that contains the code above as a standard hook.

What do you think?

**Comment by theofidry** [comment]
>So if we could somehow make it possible to hook into the post autoload dump section, we can run whatever fix we think is necessary + you could provide a utitlity that contains the code above as a standard hook.

This IMO is a very fragile hook. So I'm not too keen to provide one there as in my experience, although users may be warn this is fragile and entirely there responsibility, ultimately there still end up coming to me (not always in a friendly way) about there special case.

However your case, or rather this issue, is much more general and IMO deserves a proper fix. Be it from Composer or within PhpScoper, either way is fine by me.

So if I understand correctly, you alter the `autoload_static.php` file to change the paths?

**Comment by pxlrbt** [comment]
> So if I understand correctly, you alter the autoload_static.php file to change the paths?

It's not the paths. The hash values must be changed.

**Comment by veewee** [comment]
Thanks for the feedback! It is indeed about the hashes as @pxlrbt  mentioned.

I Agree that it is a fragile hook but on the other hand, that part of composer doesn't change that often and it gets the job done.

There aren't a lot of options:

1. Changing package names automatically inside scoper. (Also a bit edgy : See below)
2. Changing composer internal hash calculation : Not the best option as mentioned in composer/composer#7942
3. Overwriting generated autoloaders : fragile, but doesn't change that much.


Changing package names might result in unknown issues inside composer plugins.
For example for the package I am working on: it checks the name of the composer plugin to enhance the installation / deinstallation:
https://github.com/phpro/grumphp-shim/blob/master/src/Composer/GrumPHPPlugin.php#L189-L198

I can imagine that e.g. plugins like ocramius/package-versions and symfony/flex or packages that rely on those plugins, might do strange things as well.

By providing a hook, it is the package maintainer's responsibility to write something that works for him/her.

**Comment by theofidry** [comment]
>By providing a hook, it is the package maintainer's responsibility to write something that works for him/her. You can provide a hook that works and won't be that hard to maintain since the autoload files don't change that often, but it is of course provided as-is.

As mentioned, I do not want to provide an extension point when there is no use case – this issue is a use case, but one that should be fixed internally not pushed to the user via an extension point.

However this issue needs to be fixed nonetheless. And there is actually two problems: this cannot be fixed by PhpScoper out of the box in PhpScoper does not dump the autoloader (it delegates that to the user). Box however does, since it needs to be done for shipping a working PHAR.

So maybe a solution would be for PhpScoper to provide a ready to use Composer hook. When using PhpScoper directly, the user manually needs to add this hook. It is not nice but at the same time the user is already responsible for autoloading the PHP-Scoper autoloader and dumping the Composer autoloader. And within box, this hook could be appended to the scoped `composer.json` or directly change the file after dumping the autoloader – this step already exists to append the PhpScoper autoloader to the Composer autoload file

**Comment by veewee** [comment]
That solution is indeed in line with what I was trying to explain at the top of my first comment. From my point of view, that would fix it in a good enough way.

**Comment by pxlrbt** [comment]
I didn't know you can pass a callable as a composer script. That looks very clean and is way simpler than writing a composer plugin. Looks like the best way to go to include this into PHP-Scoper and mention how to use the hook in the readme.

**Comment by jarrodbell** [comment]
I ran into the same problem recently, and took the advice here to edit the `autoload_static.php` after the scoping process. I simply bolted it onto my gulp build process with a function like this (which uses the gulp-replace plugin):

```js
// Need to prefix hashes in autoload_static.php to ensure they don't conflict with any other autoloaded packages
// This is because other WordPress plugins might use same packages, scoped differently (or not scoped at all) and get loaded first (or not loaded at all)
// Which could cause our plugin to fail, or their plugins to fail. So we need to ensure each entry has a unique hash by simply adding a prefix.
const composerScopeAutoload = () => {
	return src(["path/to/vendor/composer/autoload_static.php"])
		.pipe(replace(/\'([A-Za-z0-9]*?)\' => __DIR__ \. (.*?),/g, '\'myprefix$1\' => __DIR__ . $2,'))
		.pipe(dest(`dist/vendor/composer`));
};
```

**Comment by ondrejmirtes** [comment]
In PHPStan I'm solving this manually by modifying `$GLOBALS['__composer_autoload_files']` after requiring the autoloader in the PHAR: https://github.com/phpstan/phpstan-src/blob/e8ee74f3e6c907ff5be66c917f49db637c45535b/bin/phpstan#L20-L34

**Comment by theofidry** [comment]
@ondrejmirtes do you think this somehow could be done by PHP-Scoper in `scoper-autoload.php` or this needs to be done within the same process?

**Comment by ondrejmirtes** [comment]
Maybe it could work with this logic: If the PHP-Scoper changed a file mentioned in autoload_files.php then it should change its hash...

**Comment by theofidry** [comment]
The issue is that PHP-Scoper assumes and relies on the user dumping the autoload again afterwards (hence overriding this) – You won't see that in Box because Box does it for you

**Comment by althaus** [comment]
Any momentum left here? The issue prevents us from using psalm.phar on any code using symfony/polyfill-80.

**Comment by mclaurent** [comment]
Hi

Have come across this issue aswell and awaiting a fix.

In the meantime, I've added a custom patcher that bypasses this issue.


```php
//...
return [
    'patchers' => [
        static function (string $filePath, string $prefix, string $content): string {
            $path = dirname( __FILE__ ) . DIRECTORY_SEPARATOR . implode( DIRECTORY_SEPARATOR, [ 'vendor','composer','autoload_real.php' ] );
            if ( 0 === strcasecmp( $filePath, $path ) ) {
                $content = str_replace( '\'__composer_autoload_files\'', '\'__composer_autoload_files_myapplication\'', $content );
            }
            return $content;
        },
    ],
];
````

**Comment by theofidry** [comment]
Saving this discussion with Ondřej as Slack will likely erase it soon:

- PHPStan solves it [here](https://github.com/phpstan/phpstan-src/blob/7a7b3863f91f241d5814eaa073278c2daf1fdde3/bin/phpstan#L38-L80); It unsets all hashes except for the non-prefixed ones.
- [this test](https://github.com/phpstan/phpstan-src/blob/1.9.x/tests/PHPStan/Composer/AutoloadFilesTest.php) checks to know when the autoloaded files changed

**Comment by theofidry** [comment]
Note for myself: the best lead is to update the `installed.json` file to change the name of the libraries. Alternatively it would be changing the path of the files but that looks a bit harder to me.

**Comment by theofidry** [comment]
Closed by #774

**Comment by tkfv** [comment]
hello, in a WordPress context, my plugin is scoped, the other one is not, we both depend on guzzle. With PHP scoper 0.18.3 it seems that the autoloaded guzzle file of the other plugin is still not loaded. Is this to be expected? I run `composer dump-autoload` after scoping.

**Comment by theofidry** [comment]
@tkfv would you have a reproducer by any chance?

**Comment by theofidry** [comment]
I'll be closing this issue for now. If you experience any problem please open a new issue with a reproducer so that I can investigate 🙏

**Comment by janboddez** [comment]
I seem to be running into this (well, something similar!) again. My `build/vendor/autoload_files.php` (in a WordPress plugin) looks like this:
```
<?php

// autoload_files.php @generated by Composer

$vendorDir = dirname(__DIR__);
$baseDir = dirname($vendorDir);

return array(
    '757772e28a0943a9afe83def8db95bdf' => $vendorDir . '/mf2/mf2/Mf2/Parser.php',
    'a01125dfebcda7ec3333dcd2d57ad8f2' => $baseDir . '/../includes/functions.php',
);
```
I used PHP-Scoper v0.18.7.

My `composer.json` is this:
```
{
    "require": {
        "mf2/mf2": "^0.5"
    },
    "autoload": {
        "classmap": ["includes/"],
        "files": ["includes/functions.php"]
    }
}
```

I have another plugin of which the same file looks _exactly the same_. (Because it also uses the same Parser library, and a file called `includes/functions.php`.)

When I first encountered this issue (which was fixed by moving to PHP-Scoper v0.18.x), it was the first line (`mf2/mf2/Mf2/Parser.php`) that caused problems. As if the (scoped) class was included only once, and a differently scoped (copy of that same) class would never load. Anyway, that was fixed, which is great!

This time, however, it seems to be the __second__ line (`/../includes/functions.php`).

I'm guessing this is because I used the same filename/relative path in different plugins. (The contents of both files, and their namespaces, are very different, though.) In fact, if I rename the first file (to `includes/helpers.php` or something), and then run `composer update` and again run PHP-Scoper, the conflict is gone.

But now I'm worried that any other WordPress plugin using such a filename/path (and Composer or PHP-Scoper) could cause a (future) conflict.

Or is this just how Composer works, nothing to do with PHP-Scoper?

In that case, I should probably just include the file directly (I'm going to need it anyway), and I guess I learned something today. _But then why does it work for the `Parser.php` file?_

**Comment by janboddez** [comment]
> When I first encountered this issue (which was fixed by moving to PHP-Scoper v0.18.x), it was the first line (`mf2/mf2/Mf2/Parser.php`) that caused problems.

Never mind, this thing is now also (again) failing: `Call to undefined function IndieBlocks\Mf2\parse()`

**Comment by janboddez** [comment]
And the issue seems gone now that I moved from the `Activitypub\Addon` to a `AddonForActivityPub` namespace.
